ipq40xx: Add support for MikroTik Wireless Wire dish LHGG-60ad

robimarko opened the pull request:
LHGG-60ad is IPQ4019 + wil6210 based.

Specification:

- Qualcomm IPQ4019 (717 MHz)
- 256 MB of RAM (DDR3L)
- 16 MB (SPI NOR) of flash
- 1x Gbit ethernet, 802.3af/at POE IN connected through AR8035.
- WLAN: wil6210 802.11ad PCI card
- No USB or SD card ports
- UART disabled
- 8x LEDs

Biggest news is the wil6210 PCI card.
Integration for its configuration and detection will be in other patches.

Install instructions:
Since ipq-aux-loader enables native support its installed like ar71xx RB boards.

Only current downside is that board file used is a generic one from Linux firmware and the range therefore is reduced compared to RouterOS.
This is because board file shipped in RouterOS is protected with some kind of encoding/compression that I am currently unable to figure out and revert.

Some wil6210 specifics:
CCMP is not supported, so when configuring encryption in UCI "psk+gcmp" is needed as only GCMP is supported.
Also, I have added support for 802.11ad only in UCI, there is no LuCI support so using LuCI will actually mess up your config.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
rbcfg: Enable use on Little Endian devices

This enables rbcfg to be used on modern ARM based devices
like RB450Gx4

All credits go to Sergey Sergeev <adron@mstnt.com>

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
rbcfg: Enable use on ipq40xx

Since rbcfg was updated to work on Little endian too lets enable its
use on ipq40xx.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
base-files: wifi: Add 802.11ad support

This patch adds 802.11ad support by adding 11ad hwmode identifier.
This way hostapd.sh requires no modifications as for 802.11ad only
hwmode and channel are needed as channel width is fixed to 2160MHz
and there are no HT modes.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
hostapd: wpa_supplicant: Enable proper GCMP cipher support

This patch enables hostapd.sh to properly configure wpa_supplicant
for when GCMP is used as cipher in station mode.
Without this wpa_supplicant will be unable to connect to AP.
This is needed for wil6210 as it does not support CCMP.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@chunkeey Something for you to review

xback commented:
Amazing work!
Congratz.

I have this hardware in the office and will happely help out here testing it.

robimarko commented:
Thanks, biggest help would be if someone could identify the type of compression/encoding used by Mikrotik on /bndl/wireless/lib/firmware files.
Then the performance and range would match the ROS ones

xback commented:
I'll try to pull a string at MikroTik for that info.

hauke commented:
Nice work,
60GHz is interesting.
How far did you got with extracting the board file and where can I find a dump of the flash.
Did you extract the ath10k board file from MikroTik?

robimarko commented:
Not really far, I have been trying for over a month now to somehow dump the decoded/decompressed board file from RAM but no luck, even tried using LiMe which crashed due to iomem returning impossible end address under 3.3.5 kernel.

You can find the dump here:
https://drive.google.com/file/d/1HDxf_2eJOU76LtIKhWmd4mcymdzJPK7M/view?usp=sharing
Board file for wil6210 is under /bndl/wireless/lib/firmware/wil6210-lhg-span2.8.brd but like all other 
files there, they are protected with some kind of encoding.
This is the main cause of reduced range as antenna calibration is off course not correct.

No, I did not extract any ath10k related stuff as this board does has both IPQ4019 built in radios disabled and no caldata for ath10k.
But I should have hAP AC2 soon and other developers are already working on getting caldata extracted with external Python scripts being able to do it.

xback commented:
The mail delivered 2x hap ac2 here today, and I also have some wap R ac devices 😉

robimarko commented:
Nice, some good HW.
I forgot to update my RB450Gx4 PR for a while now.

adrianschmutzler commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -85,6 +86,9 @@ ipq40xx_setup_macs()
 		lan_mac=$(macaddr_add "$wan_mac" 1)
 		;;
+	mikrotik,lhgg-60ad)
+		lan_mac=$(mtd_get_mac_binary hard_config 16)
```
Please use binary 0x10 for offset here.

robimarko replied:
Will do, I have not updated it since all offsets have been migrated to hex instead of decimal

adrianschmutzler commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4019-lhgg-60ad.dts:
```diff
@@ -0,0 +1,258 @@
+	qcom,phy_mdio_addr = <0>;
+	qcom,poll_required = <1>;
+	vlan_tag = <0 0x20>;
+};
```
Newline at the end missing.

robimarko replied:
Will fix with above

adrianschmutzler commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -14,6 +14,24 @@ define Device/Default
+		TEXT_BASE2="0x00000000" \
+		BOARD="$(BOARDNAME)" PLATFORM="$(PLATFORM)" \
+		BLOCKSIZE="$(BLOCKSIZE)" PAGESIZE="$(PAGESIZE)" \
+		UART="$(AUX_LOADER_UART)"	\
```
Any reason for the spaces before the linebreak?

robimarko replied:
Not really as far as I know.
Will fix it

adrianschmutzler commented on target/linux/ipq40xx/image/ipq-aux-loader/Makefile:
```diff
@@ -50,6 +50,7 @@ loader-compile: $(PKG_BUILD_DIR)/.prepared
 		TEXT_BASE2_FAT=$(TEXT_BASE2_FAT) 													\
 		CPU_TYPE=$(CPU_TYPE) KERNEL_IMAGE=$(LOADER_DATA)					\
 		BLOCKSIZE=$(BLOCKSIZE) PAGESIZE=$(PAGESIZE) 							\
+		UART=$(UART) 							\
```
Any reason for spaces before newline?

robimarko commented:
@adrianschmutzler Your remarks have been fixed.

adrianschmutzler commented on target/linux/ipq40xx/image/ipq-aux-loader/src/Makefile:
```diff
@@ -0,0 +1,162 @@
+			-fno-stack-protector -fstack-usage -pipe 										\
+			-marm -mno-thumb-interwork -mabi=aapcs-linux -march=armv7-a \
+			-mno-unaligned-access -fno-builtin -ffreestanding 				  \
+			-g -Os -fno-common -ffixed-r8
```
Still a lot of added spaces here.
Maybe grep your changed files with some like
grep "  \" (two spaces plus backslash)

adrianschmutzler replied:
As with the other suggestion, maybe do git format-patch on all changes first, and then only grep your patches. This will prevent a lot of results from code you are not touching.

robimarko replied:
Yeah, this is gonna be a pain to sort out.
Aux loader was not written by me so its gonna take some work to chase all of the excess spaces.

adrianschmutzler commented on target/linux/ipq40xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -93,3 +93,13 @@ platform_do_upgrade() {
+		[ -z "$(rootfs_type)" ] && mtd erase firmware
+		;;
+	esac
+}
```
Missing newline here, there are also others.
Maybe try to grep them, too (it's a little harder here, since it involves newlines).

An easy option would be to use format-patch and then grep for "No newline at end of file" in those.

robimarko replied:
Yeah, grepping in patches was actually quite fast.
This should be resolved.

robimarko committed:
linux-firmware: Add wil6210 firmware

This patch adds wil6210 firmware and board files.
Firmware version is not up to date but is only freely redistributable one I found.
Board file is a generic one so most devices and especially those for long distance
PtP links will require so in a ipq-wifi like way.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
mac80211: Add wil6210 driver

This patch adds wil6210 driver for Wilocity/QCA based 802.11ad
PCI cards.

Driver uses cfg80211 and nl80211 but not mac80211.
Integration for UCI and LuCI will come in other patches.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
ipq40xx: Add support for Mikrotik Wireless Wire dish LHGG-60ad

LHGG-60ad is IPQ4019 + wil6210 based.

Specification:

- Qualcomm IPQ4019 (717 MHz)
- 256 MB of RAM (DDR3L)
- 16 MB (SPI NOR) of flash
- 1x Gbit ethernet, 802.3af/at POE IN connected through AR8035.
- WLAN: wil6210 802.11ad PCI card
- No USB or SD card ports
- UART disabled
- 8x LEDs

Biggest news is the wil6210 PCI card.
Integration for its configuration and detection will be in other patches.

Install instructions:
Since ipq-aux-loader enables native support its installed like ar71xx RB boards.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
ipq40xx: Add ipq-aux-loader

This adds ipq-aux-loader from Sergey Sergeev <adron@yapic.net>
It is small loader for legacy or LZMA compressed FIT images.
This is needed to add support for Mikrotik RouterBoards as they
expect a specific ELF image.
This loader enables to use LZMA compressed FIT image which will
decompress the kernel image and boot it.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@adrianschmutzler I hope that I managed to resolve your remarks

adrianschmutzler commented on package/kernel/mac80211/files/lib/wifi/mac80211.sh:
```diff
@@ -76,21 +76,27 @@ detect_mac80211() {
-		ht_capab=""
-
-		iw phy "$dev" info | grep -q 'Capabilities:' && htmode=HT20
+		if ! [ iw phy "$dev" info | grep -q '58320 MHz' ]; then
```
Don't think you need the brackets here?

adrianschmutzler replied:
And is the negation (!) correct in this case?

robimarko replied:
Gotta test whether brackets are needed but ! negation is incorrect, it's leftover and I hit the case on hAP ac2 where it set ad as hwmode.
Will fix that

robimarko replied:
Looking at it again, I know why ! is used.
Its because grep -q will return 0 if match occurred, but there is also a downside that it returns 0 even if an error occurs.
So this will be changed

adrianschmutzler replied:
Can you give an example? In a quick test it works for me.

adrianschmutzler commented on package/kernel/mac80211/files/lib/wifi/mac80211.sh:
```diff
@@ -76,21 +76,27 @@ detect_mac80211() {
+			htmode=""
+			ht_capab=""
 
-		iw phy "$dev" info | grep -q '5180 MHz' && {
```
I'd personally prefer if/elif logic to the nested approach here. It won't be much difference in the end, but I think it would be more obvious to the reader. Also moved around definitions a bit:
```
htmode=""
ht_capab=""

if iw phy "$dev" info | grep -q '58320 MHz'; then
	mode_band="ad"
	channel="1"
elif iw phy "$dev" info | grep -q '5180 MHz'; then
	mode_band="a"
	channel="36"
	iw phy "$dev" info | grep -q 'Capabilities:' && htmode=HT20
	iw phy "$dev" info | grep -q 'VHT Capabilities' && htmode="VHT80"
else
	mode_band="g"
	channel="11"
	iw phy "$dev" info | grep -q 'Capabilities:' && htmode=HT20
fi

[ -n "$htmode" ] && ht_capab="set wireless.radio${devidx}.htmode=$htmode"
```

adrianschmutzler commented on package/network/services/hostapd/files/hostapd.sh:
```diff
@@ -874,6 +874,14 @@ wpa_supplicant_add_network() {
+			append network_data "pairwise=GCMP" "$N$T"
+			append network_data "group=GCMP" "$N$T"
+		;;
+		*) ;;
```
Had a similar case recently:
Is there a reason for defining an empty default case?

zorun commented:
Nice, the hap-ac2 port is based on this code, so I'm looking forward to seeing it merged!

@robimarko somebody managed to reverse the caldata encoding, AFAIR it's LZO but with a fixed part. The post is here: https://forum.openwrt.org/t/support-for-mikrotik-rb3011uias-rm/4064/425

robimarko commented:
@zorun Yeah, I saw that.
I am currently working on a C based package to extract stuff from hard_config, it will also support caldata extraction.
It's coming along, it's based on the abandoned port of routerboot.c from old ar71xx.
https://github.com/robimarko/openwrt/tree/hAP-AC2/package/utils/rbextract/src

zorun commented:
@robimarko ok, let's continue the discussion here then.

Extracting the caldata in userspace is an excellent idea, it will much easier than the routerboot.c stuff.

I have added basic support for the weird caldata extraction on hap-ac2, based on ius's python script: https://github.com/zorun/openwrt/commit/2e4099910c7784a588f17375a9e073b9610a652a

It works but here are some caveats:

- it almost certainly breaks support for some previous mikrotik devices (I don't know if you intend to use this new rbextract method for older devices but it would be nice)
- for some reason, I get a `LZO_E_INPUT_NOT_CONSUMED` error after running LZO decompression, which means that the input has too much data. Probably not that harmful.

robimarko commented:
Yeah, routerboot.c was a neat hack on ar71xx but its not really practical.
Thanks for porting the extraction for hap ac2, I bet that all MikroTik IPQ based devices use it.

And as caveats go:

- Yes, it for sure breaks backwards compatibility as caldata extraction will not work for sure and possibly part of function that locates the hard config tags.
Other stuff should work as both little and big endianess is supported.
This part is gonna need for work as I presume that ath79 is a necessity for MikroTik devices, I don't understand why there has been so little interest in it.

- This one may not be an error, here is what quick googling finds:
`LZO_E_INPUT_NOT_CONSUMED
    The end of the compressed block has been detected before all
    bytes in the compressed block have been used.
    This may actually not be an error (if 'src_len' is too large).`

ius commented:
@zorun @robimarko for the radio data tlv element (0x0016) MikroTik determines the LZO input length by discarding trailing null bytes until the LZO trailer (_11 00 00_) is found.

I.e. to find the number of padding bytes to remove, count null bytes going backwards until you hit 0x11, then subtract 2.

As for the tool itself, perhaps it's worth considering adding (partial) hcfg support to rbcfg instead? Having two userspace utilities for the ~same thing seems a bit superfluous.

robimarko commented:
@zorun @ius For whatever reason once compiled for ipq40xx rbextract will actually generate completely different file than on desktop.

I thought that maybe minilzo was the issue, so I linked in liblzo and switched to CMake, but still the same issue.
Info returned by the app is the same
Desktop:

> [robimarko@localhost rbextract]$ ./rbextract -e mtd1_hard_config.bin test.bin 
Copying fixed LZO prefix (size: 1380)
Copying input data (size: 2964)
Decompressing with LZO
Looking for ERD data in decompressed output
Found ERD magic at offset 1660
Decompress ERD data with RLE

hAP-ac2:

> root@OpenWrt:/lib/firmware/for-compare# rbextract -e /dev/mtd1 test.bin
Copying fixed LZO prefix (size: 1380)
Copying input data (size: 2964)
Decompressing with LZO
Looking for ERD data in decompressed output
Found ERD magic at offset 1660
Decompress ERD data with RLE

Any ideas?

https://github.com/robimarko/openwrt/commit/bfeb36fd82424e40fa9a81506e754b318a393d3a

robimarko commented:
There has been some progress with board files although checksum is still wrong.
https://forum.openwrt.org/t/support-for-mikrotik-rb3011uias-rm/4064/474?u=robimarko

adrianschmutzler commented on package/kernel/mac80211/files/lib/wifi/mac80211.sh:
```diff
@@ -76,21 +76,27 @@ detect_mac80211() {
-		ht_capab=""
-
-		iw phy "$dev" info | grep -q 'Capabilities:' && htmode=HT20
+		if iw phy "$dev" info | grep -cim1 '58320 MHz'; then
```
This will print the number of lines to stdout

robimarko replied:
You are correct, I have reverted to -q version but without brackets and that works.
Tested on LHGG for ad and hAP ac2 for ac.

adrianschmutzler replied:
Ah, now I understand what you did in the first place ;-)

robimarko committed:
mac80211: Add 802.11ad support

This patch adds support for 802.11ad to mac80211.sh script.
It checks whether device supports 802.11ad Channel 1 and if so
configures it to channel 1 and ad hwmode.
htmode is not needed for 802.11ad.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@adrianschmutzler Finally board file has been decrypted, so now I am wondering how to go on about it.
To modify the existing ipq-wifi package to work with wil6210 or introduce another package?

Any tips on that?

adrianschmutzler commented:
@robimarko Sorry, I cannot really help you on ipq40xx. I can only make those drive-by comments concerned with target-unspecific stuff :-)

PolynomialDivision commented:
Please merge the commits:

    linux-firmware: Add wil6210 firmware 
    mac80211: Add wil6210 driver
    mac80211: Add 802.11ad support
    base-files: wifi: Add 802.11ad support

Every 60 ghz capable device depends on it. So there can not be any progress with other 60 GHz SoCs until merge has happend. Thanks. :)

adrianschmutzler commented:
@robimarko I'd recommend to send the patches introducing 802.11ad support to the list as a separate patchset (without the device support). As pointed out by @PolynomialDivision , other devices-to-be-added might depend on it, and you will reach more people caring for such a change at the mailing list. You could use your cover letter to refer to this PR ...

adrianschmutzler commented:
@robimarko I've been so insolent to submit the 802.11ad patches to the list, of course keeping your authorship.

robimarko commented:
Just saw it, thanks for doing that.
But for me it seems that Github PRs are ignored while mailing list submissions are accepted even when not up to the target standard.
For example:
https://github.com/openwrt/openwrt/commit/affe633be4c8122835e255425418ba1680a652e3#diff-56084c5e657badcdff6ce29181444b61

Just take a look at the DTS, it was clearly just QSDK hacked.

BTW, I gotta rebase this to apply to master

adrianschmutzler commented:
@robimarko 
1. It's an open secret that certain subjects receive little (or at least less frequent) care on GitHub
2. Review standards seem to be relatively individual. Despite that, there seems to be a holdup on ipq40xx at the moment ...
3. Note that I altered patches 3 and 5 in the patchset, effectively applying the changes suggested in the comments above. Just in case you might want to add the altered patches after rebase. Patch 1 (from patchset) seems to lack proper wrapping of commit message, and all patches need to have the first word after "prefix:" in commit title in lower-case. (All addressed in patchset)

Thanks for taking this up again, I will have to rely on you anyway if there are non-trivial requests for changes.

PolynomialDivision commented:
Here is the link to the [mailing list](https://www.mail-archive.com/openwrt-devel@lists.openwrt.org/msg49313.html).

PolynomialDivision commented:
Okay. `wifi` does not work because it is calling `ubus call network.wireless up "{ }"`. Manually starting `hostapd` with the config file that is generated is working without any problems.

    daemon.notice netifd: radio0 (3167): command failed: Invalid argument (-22)
    daemon.notice netifd: radio0 (3167): command failed: Not supported (-95)
    daemon.notice netifd: radio0 (3167): command failed: Not supported (-95)
    kern.err kernel: [  263.765054] wil6210 0000:3c:00.0 wlan0: wil_cfg80211_del_iface: cannot remove the main interface
    daemon.notice netifd: radio0 (3167): command failed: Invalid argument (-22)
    daemon.notice netifd: radio0 (3167): Device setup failed: INTERFACE_CREATION_FAILED
    kern.err kernel: [  263.780825] wil6210 0000:3c:00.0 wlan0: wil_cfg80211_add_iface: virtual interfaces not supported

Further wpa_supplicant has a Segmentation fault accessing `libubus`. The only thing that helps is to compile wpa_supplicant without the `600-ubus_support.patch` patch.

    kern.info kernel: [  398.120935] wpa_supplicant[3500]: segfault at 94 ip 00007f3b6660b9c8 sp 00007ffca37c6a70 error 4 in libubus.so[7f3b6660b000+3000]
    kern.info kernel: [  398.126124] Code: ff ff 48 3d 00 00 10 00 77 4f 48 8d 43 10 48 89 6b 58 45 89 e8 44 89 f1 48 89 43 18 4c 89 e2 48 89 ef 41 83 c9 ff 48 89 43 10 <8b> 85 94 00 00 00 44 89 6b 28 8d 70 01 48 89 5b 08 48 89 1b 66 89    

My setup is `x86_64` with a QCA9008.

dangowrt commented:
While I do understand what's happening with netifd/hostapd (and that will have to be fixed, probably other FullMAC devices without support for virtual interfaces suffer from similar problems), the problem with wpa_supplicant is more of a mystery to me:
> Further, wpa_supplicant has a Segmentation fault accessing `libubus`. The only thing that helps is to compile wpa_supplicant without the `600-ubus_support.patch` patch.
> 
> ```
> kern.info kernel: [  398.120935] wpa_supplicant[3500]: segfault at 94 ip 00007f3b6660b9c8 sp 00007ffca37c6a70 error 4 in libubus.so[7f3b6660b000+3000]
> kern.info kernel: [  398.126124] Code: ff ff 48 3d 00 00 10 00 77 4f 48 8d 43 10 48 89 6b 58 45 89 e8 44 89 f1 48 89 43 18 4c 89 e2 48 89 ef 41 83 c9 ff 48 89 43 10 <8b> 85 94 00 00 00 44 89 6b 28 8d 70 01 48 89 5b 08 48 89 1b 66 89    
> ```
When does this crash occur? Upon startup of wpa_supplicant or when netifd calls wpa_supplicant.phy0 config_add ...?
Also, please share the output of `iw list` so to get an idea what could have went wrong there...

ynezz commented:
> When does this crash occur? Upon startup of wpa_supplicant or when netifd calls wpa_supplicant.phy0 config_add ...?

@dangowrt  there seems to be missing `strdup` for `blobmsg_get_string` strings, perhaps introduction of `blobmsg_get_alloc_string` would make sense?

PolynomialDivision commented:
> When does this crash occur? Upon startup of wpa_supplicant or when netifd calls wpa_supplicant.phy0 config_add ...?

I have to compile whole OpenWrt with gdb support, or? Then I can make a backtrace?
I will do this. Until I have the result, here is the `iw list`

    Wiphy phy0
    	max # scan SSIDs: 1
    	max scan IEs length: 1024 bytes
    	max # sched scan SSIDs: 16
    	max # match sets: 16
    	max # scan plans: 2
    	max scan plan interval: -1
    	max scan plan iterations: 0
    	Retry short limit: 7
    	Retry long limit: 4
    	Coverage class: 0 (up to 0m)
    	Available Antennas: TX 0 RX 0
    	Supported interface modes:
    		 * managed
    		 * AP
    		 * monitor
    		 * P2P-client
    		 * P2P-GO
    		 * P2P-device
    	Band 3:
    		Capabilities: 0x00
    			HT20
    			Static SM Power Save
    			No RX STBC
    			Max AMSDU length: 3839 bytes
    			No DSSS/CCK HT40
    		Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
    		Minimum RX AMPDU time spacing: 8 usec (0x06)
    		HT TX/RX MCS rate indexes supported: 1-12
    		Frequencies:
    			* 58320 MHz [1] (0.0 dBm)
    			* 60480 MHz [2] (0.0 dBm)
    			* 62640 MHz [3] (0.0 dBm)
    	interface combinations are not supported
    	Supported extended features:

PolynomialDivision commented:
@ynezz , @dangowrt 
> When does this crash occur? Upon startup of wpa_supplicant or when netifd calls wpa_supplicant.phy0 config_add ...?

    Starting program: /usr/sbin/wpa_supplicant -iwlan0 -c wpa.conf
    Successfully initialized wpa_supplicant
    
    Program received signal SIGSEGV, Segmentation fault.
    0x00007ffff7f719c8 in __ubus_start_request (peer=0, cmd=6, msg=0x55555566a160, req=0x7fffffffe720, ctx=0x0)
        at /XXX/nickfork/staging_dir/target-x86_64_musl/usr/include/libubox/list.h:57
    57	/XXX/nickfork/staging_dir/target-x86_64_musl/usr/include/libubox/list.h: No such file or directory.

dangowrt commented:
So apparently `wpas_ubus_add_bss()` is called before `wpas_ubus_init()`, so ctx is uninizialied. This happens because wpa_supplicant is called directly from the command line instead of using the ubus interface. I've now fixed this issue in commit 50a59b3a392aeb44c417ac3d7c2cdeea82ef2f15.
However, it should be unrelated to the segfault observed previously when using `ubus call wpa_supplicant.phy0 config_add '...'` and ubus context has certainly already been initialized at that point.

f00b4r0 commented:
For what it's worth, there seems to be a common misunderstanding about the endianness of hard_config/soft_config data on flash. It is (as one would expect) cpu-endian. There's absolutely no need for the get_32()/put_32() mumbo jumbo that's found in rbcfg and routerboot.c

I'm currently working on a clean driver to expose routerboot accessors via sysfs, I'd welcome testing on ramips/ipq if someone has the time to test an initramfs :)
When I'm done and if accepted, I expect to be able to completely rewrite rbcfg into something a lot simpler.

PolynomialDivision commented:
Did someone else tested the PR? I have several issues with the `wifi` call.

PolynomialDivision commented:
> While I do understand what's happening with netifd/hostapd (and that will have to be fixed, probably other FullMAC devices without support for virtual interfaces suffer from similar problems), the problem with wpa_supplicant is more of a mystery to me:

@dangowrt Any news regarding the FullMAC stuff? Can't we use netfid if we have no virtual interface support?

dangowrt commented:
@robimarko Please rebase and re-test with current master, so we can finally get this merged.
@PolynomialDivision I'll happily fix whatever is still broken there if I can reproduce the bug(s) somehow. I don't own any FullMAC devices in the moment, so ideally someone would have to give me remote access to one of those 11ad devices or send one over to me.

PolynomialDivision commented:
> @robimarko Please rebase and re-test with current master, so we can finally get this merged.
> @PolynomialDivision I'll happily fix whatever is still broken there if I can reproduce the bug(s) somehow. I don't own any FullMAC devices in the moment, so ideally someone would have to give me remote access to one of those 11ad devices or send one over to me.

Which kernel? 5.4 or 4.19?

dangowrt commented:
> > @robimarko Please rebase and re-test with current master, so we can finally get this merged.
> > @PolynomialDivision I'll happily fix whatever is still broken there if I can reproduce the bug(s) somehow. I don't own any FullMAC devices in the moment, so ideally someone would have to give me remote access to one of those 11ad devices or send one over to me.
> 
> Which kernel? 5.4 or 4.19?

5.4 ideally. No reason to add anything new to 4.19 imho.

PolynomialDivision commented:
> > > @robimarko Please rebase and re-test with current master, so we can finally get this merged.
> > > @PolynomialDivision I'll happily fix whatever is still broken there if I can reproduce the bug(s) somehow. I don't own any FullMAC devices in the moment, so ideally someone would have to give me remote access to one of those 11ad devices or send one over to me.
> > 
> > 
> > Which kernel? 5.4 or 4.19?
> 
> 5.4 ideally. No reason to add anything new to 4.19 imho.

I will port it to 4.19 and afterwards 5.4. Last time I tried 5.4 with this branch, my wireless wire was not starting. I have to get access to serial console again to port it to 5.4. But I will. :)

dangowrt commented:
If in the meantime you want any help debugging mac80211.sh, let me know.

dangowrt commented:
I've made some changes to supposedly improve mac80211.sh on FullMAC devices which do not support virtual interfaces.
https://github.com/openwrt/openwrt/commit/0495324b9bd0b6a976ea731a1b4b77b26df32ee6

arkap commented:
I've built an image with your changes and kernel 4.19. However, `wifi up` outputs an error:
```
daemon.notice netifd: radio0 (2120): Could not find PHY for device 'radio0'
daemon.notice netifd: radio0 (2131): WARNING: Variable 'data' does not exist or is not an array/object
```
This is my `/etc/config/wireless`:
```
config wifi-device 'radio0'
	option type 'mac80211'
	option channel '1'
	option hwmode '11ad'
	option path '/sys/devices/platform/soc/40000000.pci/pci0000:00/0000:00:00.0/0000:01:00.0'
	option disabled '0'

config wifi-iface 'default_radio0'
	option device 'radio0'
	option network 'lan'
	option mode 'ap'
	option ssid 'OpenWrt'
	option encryption 'none'
```
and the output of `iw phy`:
```
Wiphy phy0
	max # scan SSIDs: 1
	max scan IEs length: 1024 bytes
	max # sched scan SSIDs: 16
	max # match sets: 16
	max # scan plans: 2
	max scan plan interval: -1
	max scan plan iterations: 0
	Retry short limit: 7
	Retry long limit: 4
	Coverage class: 0 (up to 0m)
	Available Antennas: TX 0 RX 0
	Supported interface modes:
		 * managed
		 * AP
		 * monitor
		 * P2P-client
		 * P2P-GO
		 * P2P-device
	Band 3:
		Capabilities: 0x00
			HT20
			Static SM Power Save
			No RX STBC
			Max AMSDU length: 3839 bytes
			No DSSS/CCK HT40
		Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
		Minimum RX AMPDU time spacing: 8 usec (0x06)
		HT TX/RX MCS rate indexes supported: 1-12
		Frequencies:
			* 58320 MHz [1] (0.0 dBm)
			* 60480 MHz [2] (0.0 dBm)
			* 62640 MHz [3] (0.0 dBm)
	interface combinations are not supported
	Supported extended features:
```

dangowrt commented:
The sysfs path looks a bit odd...
> /sys/devices/platform/soc/40000000.pci/pci0000:00/0000:00:00.0/0000:01:00.0

Please explore sysfs around the pci bus as well as `/sys/class/ieee80211/phy0` which should contain a link called `device` which should match the `path` in UCI and tell us what you see :)

arkap commented:
```
> ls -la /sys/class/ieee80211/phy0/
(...)
lrwxrwxrwx    1 root     root             0 Apr 22 12:14 device -> ../../../0000:01:00.0
```
and
```
> ls -la /sys/class/ieee80211/phy0/../../../0000\:01\:00.0/
drwxr-xr-x    5 root     root             0 Jan  1  1970 .
drwxr-xr-x    7 root     root             0 Jan  1  1970 ..
-r--r--r--    1 root     root          4096 Apr 27 09:25 aer_dev_correctable
-r--r--r--    1 root     root          4096 Apr 27 09:25 aer_dev_fatal
-r--r--r--    1 root     root          4096 Apr 27 09:25 aer_dev_nonfatal
-r--r--r--    1 root     root          4096 Apr 27 09:25 ari_enabled
-rw-r--r--    1 root     root          4096 Apr 27 09:25 broken_parity_status
-r--r--r--    1 root     root          4096 Apr 27 09:25 class
-rw-r--r--    1 root     root          4096 Apr 27 09:25 config
-r--r--r--    1 root     root          4096 Apr 27 09:25 consistent_dma_mask_bits
-r--r--r--    1 root     root          4096 Apr 27 09:25 current_link_speed
-r--r--r--    1 root     root          4096 Apr 27 09:25 current_link_width
-r--r--r--    1 root     root          4096 Apr 27 09:25 device
-r--r--r--    1 root     root          4096 Apr 27 09:25 devspec
-r--r--r--    1 root     root          4096 Apr 27 09:25 dma_mask_bits
lrwxrwxrwx    1 root     root             0 Apr 27 09:25 driver -> ../../../../../../../bus/pci/drivers/wil6210
-rw-r--r--    1 root     root          4096 Apr 27 09:25 driver_override
-rw-r--r--    1 root     root          4096 Apr 27 09:25 enable
drwxr-xr-x    3 root     root             0 Apr 22 12:14 ieee80211
-r--r--r--    1 root     root          4096 Apr 27 09:25 irq
-r--r--r--    1 root     root          4096 Apr 27 09:25 local_cpulist
-r--r--r--    1 root     root          4096 Apr 27 09:25 local_cpus
-r--r--r--    1 root     root          4096 Apr 27 09:25 max_link_speed
-r--r--r--    1 root     root          4096 Apr 27 09:25 max_link_width
-r--r--r--    1 root     root          4096 Apr 27 09:25 modalias
-rw-r--r--    1 root     root          4096 Apr 27 09:25 msi_bus
drwxr-xr-x    2 root     root             0 Apr 27 09:25 msi_irqs
drwxr-xr-x    3 root     root             0 Apr 27 09:25 net
lrwxrwxrwx    1 root     root             0 Apr 27 09:25 of_node -> ../../../../../../../firmware/devicetree/base/soc/pci@40000000/bridge@0,0/wifi@1,0
--w--w----    1 root     root          4096 Apr 27 09:25 remove
--w--w----    1 root     root          4096 Apr 27 09:25 rescan
--w-------    1 root     root          4096 Apr 27 09:25 reset
-r--r--r--    1 root     root          4096 Apr 27 09:25 resource
-rw-------    1 root     root       2097152 Apr 27 09:25 resource0
-r--r--r--    1 root     root          4096 Apr 27 09:25 revision
lrwxrwxrwx    1 root     root             0 Apr 27 09:25 subsystem -> ../../../../../../../bus/pci
-r--r--r--    1 root     root          4096 Apr 27 09:25 subsystem_device
-r--r--r--    1 root     root          4096 Apr 27 09:25 subsystem_vendor
-rw-r--r--    1 root     root          4096 Apr 27 09:25 uevent
-r--r--r--    1 root     root          4096 Apr 27 09:25 vendor
```

I've changed the line in `/etc/config/wireless` to `option path 'pci0000:00/0000:00:00.0/0000:01:00.0'` and now everything works as expected.
So it seems there is an issue in the detection of the PCI path of the phy.

dangowrt commented:
```
ls -la /sys/class/ieee80211/phy0/
(...)
lrwxrwxrwx    1 root     root             0 Apr 22 12:14 device -> ../../../0000:01:00.0
```
Apparentlty `mac80211_phy_to_path` in `/lib/netifd/wireless/mac80211.sh` isn't ok with handling this kinda weird relative path in the symlink...
@nbd168 added this function in commit 8b15e7f661b not very long ago, maybe he can take a look at the situation here.

arkap commented:
```
root@owrt-ww-01:~# . /lib/netifd/wireless/mac80211.sh 
root@owrt-ww-01:~# mac80211_phy_to_path phy0
soc/40000000.pci/pci0000:00/0000:00:00.0/0000:01:00.0
```

f00b4r0 commented:
Please note that a number of changes have just been merged in master (from #2943) that impact all mikrotik devices in ramips/ath79. It would probably be good if all new mikrotik devices built upon these changes going forward. HTH

PolynomialDivision commented:
> Please note that a number of changes have just been merged in master (from #2943) that impact all mikrotik devices in ramips/ath79. It would probably be good if all new mikrotik devices built upon these changes going forward. HTH

Since this PR is for `ipq40xx` the pachtes you mentioned are not important?

robimarko commented:
> > Please note that a number of changes have just been merged in master (from #2943) that impact all mikrotik devices in ramips/ath79. It would probably be good if all new mikrotik devices built upon these changes going forward. HTH
> 
> Since this PR is for `ipq40xx` the pachtes you mentioned are not important?

They still matter due to partition parser and sysfs driver for hard_config.

I have started rebasing this onto master and should have an updated version next week

xback commented:
> > > Please note that a number of changes have just been merged in master (from #2943) that impact all mikrotik devices in ramips/ath79. It would probably be good if all new mikrotik devices built upon these changes going forward. HTH
> > 
> > 
> > Since this PR is for `ipq40xx` the pachtes you mentioned are not important?
> 
> They still matter due to partition parser and sysfs driver for hard_config.
> 
> I have started rebasing this onto master and should have an updated version next week

Feel free to ping me when you need a hand in detailed testing.

arkap commented:
Has anyone made any progress with this? I've rebased @robimarko 's fork onto master today but after a sysupgrade, the device doesn't boot right. I'd need to solder some serial header onto the board to see what's going on.

adrianschmutzler commented:
This PR is open for more than 9 months now, and nobody actually showed interest in reviewing or merging it; same appears for the 802.11ad patches I put on openwrt-devel list separately. Since I don't have any reason to expect this condition to change, I'm closing this now. Sorry.

dangowrt commented:
I was under the impression that there was some traction lately when it comes to adding MikroTik stuff and I do believe there is interest in this hardware, especially as people are posting reports of tests just minutes before.

adrianschmutzler commented:
> I was under the impression that there was some traction lately when it comes to adding MikroTik stuff and I do believe there is interest in this hardware, especially as people are posting reports of tests just minutes before.

There is a lot of interest, but only by the users, not by the reviewers ... ;-)

f00b4r0 commented:
There is interest on my side, just complete lack of time right now.
Also, as I pointed out in #3037 the ipq-aux-loader this relies upon is very "questionable" to say the least.

In general with MirkroTik devices, given the apparent limitations of the TFTP loader it appears we will have to move toward a situation where: 1) the (install) initramfs is massively stripped down (both userspace and kernel) compared to the (regular) sysupgrade and 2) the sysupgrade kernel should ideally be a (possibly natively compressed) plain ELF kernel, without any extra loader bits which introduce other problems as evidenced in #3026.

Item 1) will require some rework of the build system that will have to be made in a way acceptable for OpenWRT (currently the build system does not allow building initramfs images that differ from sysupgrade, let alone using a separate kernel config there) and 2) needs investigation and testing to prove the idea is sound (my understanding is that the limitations imparted on TFTP-booted kernels do not apply to flash-booted ones, this needs to be confirmed). Both of which require significant time 😛 

Last but not least, the issue with CONFIG_MTD_SPI_NOR_USE_4K_SECTORS_LIMIT (which appears to be completely broken in recent kernels) must be resolved.

See useful pointers in #3026 (which is rather long): https://github.com/openwrt/openwrt/pull/3026#issuecomment-642901344 and https://github.com/openwrt/openwrt/pull/3026#issuecomment-640202098 (and comments onwards from there for the issues related to hitting routerboot TFTP boot limitations).

Nearly all comments made in #3037 apply to this PR as well, so worth reading.

As regards this particular PR: among other things, rbcfg is gone: it needs to be updated to use the new sysfs interface.

HTH, that's the best I can manage right now.

SvenRoederer commented:
As the is some confusion on the MikroTik productnames, I opened #3270, where we can probably agree on a guideline.

dangowrt commented:
@xback @f00b4r0 can anyone try if caldata gets available in sysfs when using
https://patchwork.ozlabs.org/project/openwrt/patch/20200824103840.38920-1-hacks@slashdirt.org/

lynxis commented:
I've pulled this PR into my staging tree.

john-tho commented:
Hi @lynxis, thanks.
Looking at your staging branch https://git.openwrt.org/?p=openwrt/staging/lynxis.git;a=commitdiff;h=e3a58ddb17b252f948445ff4556a8a07763615c0 I saw some superficial issues:
- old loader still mentioned in commit message
- model name in LED labels
- DTS very similar to other ipq40xx mikrotik devices. Do we want to abstract out the common parts?
- DTS first partition has wrong address `0` in label `partition0@0`. It looks to me like the only difference with SPI-NOR to the hap ac2 are the GPIOs used. Should be able to use the hap ac2 partition node which uses the routerboot parser if wanted.
 
Cheers

dangowrt commented:
@john-tho @lynxis @robimarko I've pulled the commit adding the device from lynxis tree and updated it. I can only compile test it (I don't have the hardware), so it'd be great if someone with the hardware could give it a shot. Then we can finally get this merged.

https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=commit;h=145b4dedc3f7cc73abf3f8c6fdeb1df233d106f0

PolynomialDivision commented:
@dangowrt I can test it, but without 60 ghz wifi p2p connection. I only have one dish available.

PolynomialDivision commented:
I saw some mistakes in the device tree. I shortly fix it.

PolynomialDivision commented:
@dangowrt I fixed the ethernet interface:
https://github.com/PolynomialDivision/openwrt/commit/becde4ce1aaab466f0f892266ab140092e792a2a

PolynomialDivision commented:
```
root@OpenWrt:~# dmesg
[    0.000000] Booting Linux on physical CPU 0x0
[    0.000000] Linux version 5.10.80 (nick@cool) (arm-openwrt-linux-muslgnueabi-gcc (OpenWrt GCC 11.2.0 r18107+28-db34b93331) 11.2.0, GNU ld (GNU Binutils) 2.37) #0 SMP Fri Nov 19 19:30:23 2021
[    0.000000] CPU: ARMv7 Processor [410fc075] revision 5 (ARMv7), cr=10c5387d
[    0.000000] CPU: div instructions available: patching division code
[    0.000000] CPU: PIPT / VIPT nonaliasing data cache, VIPT aliasing instruction cache
[    0.000000] OF: fdt: Machine model: Mikrotik Wireless Wire Dish LHGG-60ad
[    0.000000] Memory policy: Data cache writealloc
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000080000000-0x000000008fffffff]
[    0.000000]   HighMem  empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000080000000-0x0000000087dfffff]
[    0.000000]   node   0: [mem 0x0000000087e00000-0x0000000087ffffff]
[    0.000000]   node   0: [mem 0x0000000088000000-0x000000008fffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000080000000-0x000000008fffffff]
[    0.000000] On node 0 totalpages: 65536
[    0.000000]   Normal zone: 576 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 65536 pages, LIFO batch:15
[    0.000000] percpu: Embedded 15 pages/cpu s30732 r8192 d22516 u61440
[    0.000000] pcpu-alloc: s30732 r8192 d22516 u61440 alloc=15*4096
[    0.000000] pcpu-alloc: [0] 0 [0] 1 [0] 2 [0] 3 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 64960
[    0.000000] Kernel command line: 
[    0.000000] Dentry cache hash table entries: 32768 (order: 5, 131072 bytes, linear)
[    0.000000] Inode-cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 239008K/262144K available (6141K kernel code, 602K rwdata, 1548K rodata, 9216K init, 236K bss, 23136K reserved, 0K cma-reserved, 0K highmem)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000] 	Tracing variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 10 jiffies.
[    0.000000] NR_IRQS: 16, nr_irqs: 16, preallocated irqs: 16
[    0.000000] random: get_random_bytes called from start_kernel+0x360/0x508 with crng_init=0
[    0.000000] arch_timer: cp15 timer(s) running at 48.00MHz (virt).
[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0xb11fd3bfb, max_idle_ns: 440795203732 ns
[    0.000008] sched_clock: 56 bits at 48MHz, resolution 20ns, wraps every 4398046511096ns
[    0.000024] Switching to timer-based delay loop, resolution 20ns
[    0.000230] Calibrating delay loop (skipped), value calculated using timer frequency.. 96.00 BogoMIPS (lpj=480000)
[    0.000254] pid_max: default: 32768 minimum: 301
[    0.000444] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.000467] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.001527] CPU: Testing write buffer coherency: ok
[    0.001849] qcom_scm: convention: smc legacy
[    0.002632] Setting up static identity map for 0x80300000 - 0x8030003c
[    0.002771] rcu: Hierarchical SRCU implementation.
[    0.002979] dyndbg: Ignore empty _ddebug table in a CONFIG_DYNAMIC_DEBUG_CORE build
[    0.003250] smp: Bringing up secondary CPUs ...
[    0.006405] smp: Brought up 1 node, 4 CPUs
[    0.006431] SMP: Total of 4 processors activated (384.00 BogoMIPS).
[    0.006443] CPU: All CPU(s) started in SVC mode.
[    0.011045] VFP support v0.3: implementor 41 architecture 2 part 30 variant 7 rev 5
[    0.011211] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.011240] futex hash table entries: 1024 (order: 4, 65536 bytes, linear)
[    0.011517] pinctrl core: initialized pinctrl subsystem
[    0.012738] NET: Registered protocol family 16
[    0.013147] DMA: preallocated 256 KiB pool for atomic coherent allocations
[    0.014322] thermal_sys: Registered thermal governor 'step_wise'
[    0.014740] cpuidle: using governor ladder
[    0.014799] cpuidle: using governor menu
[    0.035585] cryptd: max_cpu_qlen set to 1000
[    0.039302] usbcore: registered new interface driver usbfs
[    0.039379] usbcore: registered new interface driver hub
[    0.039446] usbcore: registered new device driver usb
[    0.039497] pps_core: LinuxPPS API ver. 1 registered
[    0.039509] pps_core: Software ver. 5.3.6 - Copyright 2005-2007 Rodolfo Giometti <giometti@linux.it>
[    0.039536] PTP clock support registered
[    0.041239] clocksource: Switched to clocksource arch_sys_counter
[    0.042196] NET: Registered protocol family 2
[    0.042365] IP idents hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.042989] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 6144 bytes, linear)
[    0.043047] TCP established hash table entries: 2048 (order: 1, 8192 bytes, linear)
[    0.043087] TCP bind hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.043138] TCP: Hash tables configured (established 2048 bind 2048)
[    0.043275] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.043318] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.043565] NET: Registered protocol family 1
[    0.043610] PCI: CLS 0 bytes, default 64
[    0.151947] workingset: timestamp_bits=14 max_order=16 bucket_order=2
[    0.156438] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.156463] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.292517] qcom-pcie 40000000.pci: host bridge /soc/pci@40000000 ranges:
[    0.292585] qcom-pcie 40000000.pci:       IO 0x0040200000..0x00402fffff -> 0x0040200000
[    0.292628] qcom-pcie 40000000.pci:      MEM 0x0040300000..0x0040ffffff -> 0x0040300000
[    0.572814] qcom-pcie 40000000.pci: Link up
[    0.572992] qcom-pcie 40000000.pci: PCI host bridge to bus 0000:00
[    0.573017] pci_bus 0000:00: root bus resource [bus 00-ff]
[    0.573038] pci_bus 0000:00: root bus resource [io  0x0000-0xfffff] (bus address [0x40200000-0x402fffff])
[    0.573053] pci_bus 0000:00: root bus resource [mem 0x40300000-0x40ffffff]
[    0.573106] pci 0000:00:00.0: [17cb:1001] type 01 class 0x060400
[    0.573137] pci 0000:00:00.0: reg 0x10: [mem 0x00000000-0x00000fff]
[    0.573206] pci 0000:00:00.0: PME# supported from D0 D3hot
[    0.574637] PCI: bus0: Fast back to back transfers disabled
[    0.574916] pci 0000:01:00.0: [1ae9:0310] type 00 class 0x028000
[    0.575062] pci 0000:01:00.0: reg 0x10: [mem 0x00000000-0x001fffff 64bit]
[    0.577135] PCI: bus1: Fast back to back transfers disabled
[    0.577191] pci 0000:00:00.0: BAR 8: assigned [mem 0x40400000-0x405fffff]
[    0.577211] pci 0000:00:00.0: BAR 0: assigned [mem 0x40300000-0x40300fff]
[    0.577236] pci 0000:01:00.0: BAR 0: assigned [mem 0x40400000-0x405fffff 64bit]
[    0.577308] pci 0000:00:00.0: PCI bridge to [bus 01-ff]
[    0.577325] pci 0000:00:00.0:   bridge window [mem 0x40400000-0x405fffff]
[    0.580168] pcieport 0000:00:00.0: AER: enabled with IRQ 68
[    0.582296] bam-dma-engine 8e04000.dma: num-channels unspecified in dt
[    0.582321] bam-dma-engine 8e04000.dma: num-ees unspecified in dt
[    0.583001] tcsr 1949000.tcsr: setting wifi_glb_cfg = 41000000
[    0.583100] tcsr 1953000.ess_tcsr: setting ess interface select = 1
[    0.583195] tcsr 1957000.tcsr: setting wifi_noc_memtype_m0_m2 = 2222222
[    0.583492] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.584095] msm_serial 78af000.serial: msm_serial: detected port #0
[    0.584142] msm_serial 78af000.serial: uartclk = 1843200
[    0.584199] 78af000.serial: ttyMSM0 at MMIO 0x78af000 (irq = 33, base_baud = 115200) is a MSM
[    0.584228] msm_serial: console setup on port #0
[    1.255611] printk: console [ttyMSM0] enabled
[    1.260788] msm_serial: driver initialized
[    1.269914] loop: module loaded
[    1.270970] spi_qup 78b5000.spi: IN:block:16, fifo:64, OUT:block:16, fifo:64
[    1.274025] spi-nor spi0.0: w25q128jv (16384 Kbytes)
[    1.279213] 3 fixed-partitions partitions found on MTD device spi0.0
[    1.284165] Creating 3 MTD partitions on "spi0.0":
[    1.290453] 0x000000000000-0x000000080000 : "Qualcomm"
[    1.295700] 0x000000080000-0x000000100000 : "RouterBoot"
[    1.307962] 3 routerbootpart partitions found on MTD device RouterBoot
[    1.307997] Creating 3 MTD partitions on "RouterBoot":
[    1.313421] 0x00000000e000-0x000000010000 : "hard_config"
[    1.319069] 0x000000010000-0x000000017c31 : "dtb_config"
[    1.324536] 0x00000003d000-0x00000003e000 : "soft_config"
[    1.329852] 0x000000100000-0x000001000000 : "firmware"
[    1.338993] libphy: Fixed MDIO Bus: probed
[    1.340419] libphy: ipq4019_mdio: probed
[    1.501302] EDMA using MAC@ - using
[    1.501321] 82:52:96:43:78:16
[    1.583764] i2c /dev entries driver
[    1.585592] sdhci: Secure Digital Host Controller Interface driver
[    1.586067] sdhci: Copyright(c) Pierre Ossman
[    1.592390] sdhci-pltfm: SDHCI platform and OF driver helper
[    1.603477] MikroTik RouterBOARD hardware configuration sysfs driver v0.06
[    1.605430] MikroTik RouterBOARD software configuration sysfs driver v0.03
[    1.610244] NET: Registered protocol family 10
[    1.617714] Segment Routing with IPv6
[    1.620658] NET: Registered protocol family 17
[    1.624397] 8021q: 802.1Q VLAN Support v1.8
[    1.628657] Registering SWP/SWPB emulation handler
[    1.684046] Freeing unused kernel memory: 9216K
[    1.734777] Run /init as init process
[    1.735005]   with arguments:
[    1.735085]     /init
[    1.735153]   with environment:
[    1.735225]     HOME=/
[    1.735297]     TERM=linux
[    2.075890] init: Console is alive
[    2.076215] init: - watchdog -
[    2.087492] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    2.089855] genirq: irq_chip msmgpio did not update eff. affinity mask of irq 69
[    2.108241] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    2.109444] init: - preinit -
[    2.389491] random: jshn: uninitialized urandom read (4 bytes read)
[    2.445218] random: jshn: uninitialized urandom read (4 bytes read)
[    2.488539] random: jshn: uninitialized urandom read (4 bytes read)
[    2.734124] ess_edma c080000.edma eth0: Link is Down
[    4.793440] ess_edma c080000.edma eth0: Link is Up - 100Mbps/Full - flow control off
[    4.794268] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[    6.956737] ess_edma c080000.edma eth0: Link is Down
[    6.965379] procd: - early -
[    6.965647] procd: - watchdog -
[    7.515273] procd: - watchdog -
[    7.519246] procd: - ubus -
[    7.546117] urandom_read: 5 callbacks suppressed
[    7.546131] random: ubusd: uninitialized urandom read (4 bytes read)
[    7.570571] random: ubusd: uninitialized urandom read (4 bytes read)
[    7.574308] procd: - init -
[    7.967529] kmodloader: loading kernel modules from /etc/modules.d/*
[    7.982516] Loading modules backported from Linux version v5.15-rc6-0-g519d81956ee2
[    7.982566] Backport generated by backports.git v5.15-rc6-1-0-gd44432d6
[    7.991566] urngd: v1.0.2 started.
[    8.003428] xt_time: kernel timezone is -0000
[    8.024821] PPP generic driver version 2.4.2
[    8.025998] NET: Registered protocol family 24
[    8.038124] wil6210 0000:01:00.0: wil6210 device found [1ae9:0310] (rev 2) bar size 0x200000
[    8.038724] wil6210 0000:01:00.0: enabling device (0140 -> 0142)
[    8.045789] wil6210 0000:01:00.0 (unnamed net_device) (uninitialized): wil_pcie_probe: CSR at [mem 0x40400000-0x405fffff 64bit] -> 0x(ptrval)
[    8.051713] wil6210 0000:01:00.0 (unnamed net_device) (uninitialized): wil_set_capabilities: Board hardware is Sparrow B0, flash exist
[    8.064249] wil6210 0000:01:00.0 (unnamed net_device) (uninitialized): wil_set_capabilities: platform_capa 0x0
[    8.078767] random: crng init done
[    8.080899] wil6210 0000:01:00.0 (unnamed net_device) (uninitialized): wil_refresh_fw_capabilities: keep_radio_on_during_sleep (0)
[    8.089586] wil6210 0000:01:00.0: using dma mask 48
[    8.181325] wil6210 0000:01:00.0 (unnamed net_device) (uninitialized): wil_get_bl_info: Boot Loader struct v2: MAC = 08:c5:e1:da:6b:5b RF = 0x0000 (status 0x0000) bband = 0x00000000
[    8.181391] wil6210 0000:01:00.0 (unnamed net_device) (uninitialized): wil_get_bl_info: Boot Loader build 255.255.0.6836
[    8.196516] wil6210 0000:01:00.0 (unnamed net_device) (uninitialized): wil_set_oob_mode: oob_mode to 0
[    8.220297] kmodloader: done loading kernel modules from /etc/modules.d/*
[   24.204650] br-lan: port 1(eth0) entered blocking state
[   24.204746] br-lan: port 1(eth0) entered disabled state
[   24.209586] device eth0 entered promiscuous mode
[   26.313535] ess_edma c080000.edma eth0: Link is Up - 100Mbps/Full - flow control off
[   26.314270] br-lan: port 1(eth0) entered blocking state
[   26.320640] br-lan: port 1(eth0) entered forwarding state
[   26.330015] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[  298.808410] wil6210 0000:01:00.0 wlan0: wil_cfg80211_del_iface: cannot remove the main interface
[  299.149226] wil6210 0000:01:00.0 wlan0: wil_cfg80211_add_iface: virtual interfaces not supported
[  299.213713] wil6210 0000:01:00.0 wlan0: _wil6210_disconnect: disconnect bssid=(null), reason=3
[  299.213844] wil6210 0000:01:00.0 wlan0: _wil6210_disconnect_complete: disconnect_complete: bssid=(null), reason=3
[  299.272028] wil6210 0000:01:00.0 wlan0: wil_get_bl_info: Boot Loader struct v2: MAC = 08:c5:e1:da:6b:5b RF = 0x0000 (status 0x0000) bband = 0x00000000
[  299.272157] wil6210 0000:01:00.0 wlan0: wil_get_bl_info: Boot Loader build 255.255.0.6836
[  299.284776] wil6210 0000:01:00.0 wlan0: wil_set_oob_mode: oob_mode to 0
[  299.292841] wil6210 0000:01:00.0 wlan0: wil_reset: Use firmware <wil6210.fw> + board <wil6210.brd>
[  299.522276] wil6210 0000:01:00.0 wlan0: wmi_evt_ready: FW ver. 5.2.0.18(SW 18); MAC 08:c5:e1:da:6b:5b; 0 MID's
[  299.522484] wil6210 0000:01:00.0 wlan0: wil_wait_for_fw_ready: FW ready after 210 ms. HW version 0x00000001
[  299.531754] wil6210 0000:01:00.0 wlan0: wil_configure_interrupt_moderation: set ITR_TX_CNT_TRSH = 500 usec
[  299.540883] wil6210 0000:01:00.0 wlan0: wil_configure_interrupt_moderation: set ITR_TX_IDL_CNT_TRSH = 13 usec
[  299.550661] wil6210 0000:01:00.0 wlan0: wil_configure_interrupt_moderation: set ITR_RX_CNT_TRSH = 500 usec
[  299.560587] wil6210 0000:01:00.0 wlan0: wil_configure_interrupt_moderation: set ITR_RX_IDL_CNT_TRSH = 13 usec
[  299.570156] wil6210 0000:01:00.0 wlan0: wil_refresh_fw_capabilities: keep_radio_on_during_sleep (0)
[  299.588436] br-lan: port 2(wlan0) entered blocking state
[  299.588934] br-lan: port 2(wlan0) entered disabled state
[  299.594966] device wlan0 entered promiscuous mode
[  299.599981] br-lan: port 2(wlan0) entered blocking state
[  299.604449] br-lan: port 2(wlan0) entered forwarding state
[  299.663537] device wlan0 left promiscuous mode
[  299.664869] br-lan: port 2(wlan0) entered disabled state
[  299.747584] wil6210 0000:01:00.0 wlan0: _wil6210_disconnect: disconnect bssid=(null), reason=3
[  299.748017] wil6210 0000:01:00.0 wlan0: _wil6210_disconnect_complete: disconnect_complete: bssid=(null), reason=3
[  299.814307] wil6210 0000:01:00.0 wlan0: wil_get_bl_info: Boot Loader struct v2: MAC = 08:c5:e1:da:6b:5b RF = 0x0000 (status 0x0000) bband = 0x00000000
[  299.814383] wil6210 0000:01:00.0 wlan0: wil_get_bl_info: Boot Loader build 255.255.0.6836
[  299.826904] wil6210 0000:01:00.0 wlan0: wil_set_oob_mode: oob_mode to 0
```

PolynomialDivision commented:
```
root@OpenWrt:~# iw phy
Wiphy phy0
	wiphy index: 0
	max # scan SSIDs: 1
	max scan IEs length: 1024 bytes
	max # sched scan SSIDs: 16
	max # match sets: 16
	Retry short limit: 7
	Retry long limit: 4
	Coverage class: 0 (up to 0m)
	Available Antennas: TX 0 RX 0
	Supported interface modes:
		 * managed
		 * AP
		 * monitor
		 * P2P-client
		 * P2P-GO
		 * P2P-device
	Band 3:
		Capabilities: 0x00
			HT20
			Static SM Power Save
			No RX STBC
			Max AMSDU length: 3839 bytes
			No DSSS/CCK HT40
		Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
		Minimum RX AMPDU time spacing: 8 usec (0x06)
		HT TX/RX MCS rate indexes supported: 1-12
		Frequencies:
			* 58320 MHz [1] (0.0 dBm)
			* 60480 MHz [2] (0.0 dBm)
			* 62640 MHz [3] (0.0 dBm)
	interface combinations are not supported
	max # scan plans: 2
	max scan plan interval: -1
	max scan plan iterations: 0
	Supported extended features:
```

PolynomialDivision commented:
I can not tell if the wifi is working, since I only have 1 device.

PolynomialDivision commented:
Whups, I did not commit my changes. Here is the correct device tree:
https://github.com/PolynomialDivision/openwrt/commit/14cabc372020056183288598217c8565599a4b3c

xback commented:
Hi all,
I finally have some time to work on OpenWRT again.
I would really like to merge this work.

I have a few of these laying around in the stock over here.
If this could be bumped to latest master, I'm more than willing to aid in getting it merged

xback commented:
Ahha?

https://github.com/openwrt/openwrt/commit/8d623047851337c4507f77b51f47851988c0f53c

Anything lacking to it today?

dangowrt commented:
Looks like we still need to extract/decode the boardfile for wil6210 as with the defaults range is very limited and much worse than with the vendor firmware.

f00b4r0 commented:
Wouldn't #4679 help ?

robimarko commented:
Actually not, the wil6210 board files are shipped as part of ROS along with other firmware.
But they are encrypted and compressed, I remember somebody reversing the algorithm and having a script to decompress.

This should be it, there is an explanation on the seed as well.
https://forum.openwrt.org/t/support-for-mikrotik-rb3011uias-rm/4064/474?u=robimarko

aparcar commented:
Hi, I got the device and would like to review this PR. What's currently missing? How is the firmware decryption incorporated? Looks like we dropped the `rbcfg` package a while ago, do we need it for this device?

robimarko commented:
There is no FW decryption, BDF must be manually pulled out and decrypted.
rbcfg is completely not needed but this PR is ancient at this point.

aparcar commented:
Everything but the `rbcfg` stuff seems to be already upstream. I guess this PR could be closed as it's technically merged.

Is it realistic to incorporate the decryption stuff into OpenWrt upstream? Is there a manual on what steps are required?

robimarko commented:
Not really, its completely manual, I have linked the forum post:
https://forum.openwrt.org/t/support-for-mikrotik-rb3011uias-rm/4064/474

Its one-off and that BDF can be included in OpenWrt but as I far as I can tell the header was completely missing or something like that.

aparcar commented:
I'll try to make sense out of it, I'm still in the blue where to put the decrypted files and what's the BDF stuff. First time touching this device/target.

robimarko commented:
It goes in the wil6210 FW folder like for other drivers, it's just a BDF like for ath10k, etc.
But it work work, I gave up on this a long time ago.

aparcar commented:
I'd close this PR since all commits were merged (except `rbcfg`), the improvement of `wil6210` should happen in a new PR. @robimarko are you okay with that?

robimarko commented:
Yeah, since it was merged already.
